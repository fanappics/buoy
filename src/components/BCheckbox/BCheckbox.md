Buoy Checkbox Component

### Single

```js
  const checkbox1 = {
    checkboxes:[{ id: 'checkbox-1', value: 'Sign up now' }]
  }
  <b-checkbox v-bind="checkbox1" />
```

### Multiple

```js
  const checkbox2 = {
    checkboxes:[
      { id: 'checkbox-2', value: 'Option #1' },
      { id: 'checkbox-3', value: 'Option #2' }
    ],
    groupId: 'checkbox-group-1',
    groupLabel: 'You must choose:'
  }
  <b-checkbox v-bind="checkbox2" />
```

### Required

```js
  const checkbox3 = {
    checkboxes:[
      { id: 'checkbox-4', value: 'Option #1' },
      { id: 'checkbox-5', value: 'Option #2' }
    ],
    groupId: 'checkbox-group-2',
    groupLabel: 'You must choose:'
  }
  <b-checkbox v-bind="checkbox3" required />
```

### Validation with BValidator

```js
  let checkboxGroupsValid = false;
  const checkbox4 = {
    checkboxes:[
      { id: 'checkbox-6', value: 'Winter' },
      { id: 'checkbox-7', value: 'Spring' },
      { id: 'checkbox-8', value: 'Summer' },
      { id: 'checkbox-9', value: 'Fall' }

    ],
    groupId: 'checkbox-group-3',
    groupLabel: 'I Love:'
  }

  let checked = ['checkbox-6'];
  <div>
    <b-validator scope="checkbox-group-validator" v-model="checkboxGroupsValid" >
      <b-checkbox v-bind="checkbox4" v-model="checked" required />
      <b-button :disabled="!checkboxGroupsValid" label="Submit"> />
    </b-validator>

   checked {{ checked }}
  </div>
```

### Disabled

```js
  const checkbox5 = {
    checkboxes:[
      { id: 'checkbox-10', value: 'Option #1' },
      { id: 'checkbox-11', value: 'Option #2' },
      { id: 'checkbox-12', value: 'Option #3' }
    ],
    groupId: 'checkbox-group-4',
    groupLabel: 'Select an option:'
  }
  let checked = ['checkbox-11'];
  <b-checkbox v-bind="checkbox5" disabled />
```
