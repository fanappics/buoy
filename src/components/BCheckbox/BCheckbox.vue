<template>
  <!-- Group checkboxes: requires group id and label -->
  <div v-if="groupId && groupLabel" class="column">
    <div role="group" :aria-labelledby="groupId" class="flex">
      <label :id="groupId" :class="{ bold: true, disabled: disabled }">
        {{ groupLabel }}<span v-if="required" aria-hidden="true">*</span>
      </label>
      <div class="flex wrap">
        <div v-for="(checkbox, index) in checkboxes" :key="index" class="flex baseline">
          <input type="checkbox" 
            v-model="publicValues" 
            v-validate:publicValues.initial="validations"
            v-bind="validationAttributes"
            :aria-checked="values.indexOf(checkbox.id) > -1"
            :aria-describedby="invalid ? `error-${groupId}` : ''"
            :aria-labelledby="`${groupId} label-${checkbox.id}`"
            :disabled="checkbox.disabled || disabled"
            :id="checkbox.id"
            :value="checkbox.id"
          />
          <label :id="`label-${checkbox.id}`" :for="checkbox.id">
            {{ checkbox.value }}
          </label>
        </div>
      </div>
    </div>
    <div v-if="touched && invalid" :id="`error-${groupId}`" class="error-text">
      {{ errors.first(validationId, parentScope || validationId) }}
    </div>
  </div>
  <!-- Single checkbox -->
  <div v-else-if="checkboxes.length == 1" class="column">
    <template v-for="(checkbox,index) in checkboxes">
      <div class="flex baseline" :key="index">
        <input type="checkbox" 
          v-model="publicValues" 
          v-validate:publicValues.initial="validations" 
          v-bind="validationAttributes"
          :aria-checked="values.indexOf(checkbox.id) > -1"
          :aria-describedby="invalid ? `error-${checkbox.id}` : ''"
          :aria-labelledby="`label-${checkbox.id}`"
          :disabled="checkbox.disabled || disabled"
          :id="checkbox.id" 
          :value="checkbox.id" 
          :key="index"
        />
        <label :id="`label-${checkbox.id}`" :for="checkbox.id" class="bold" :key="index">
          {{ checkbox.value }}<span v-if="required" aria-hidden="true">*</span>
        </label>
      </div>
      <div v-if="touched && invalid" :id="`error-${checkbox.id}`" class="error-text" :key="index">
        {{ errors.first(validationId, parentScope || validationId) }}
      </div>
    </template>
  </div>
</template>

<script>
import validationMixIn from '../../mixins/validation'

export default {
  name: 'b-checkbox',
  mixins: [validationMixIn()],
  props: {
    //Required props
    checkboxes: {
      type: Array,
      required: true,
      validator (value) {
        if (value.length < 1)
          throw new TypeError('No checkboxes were defined in the array')

        for (let checkbox of value) {
          if (!checkbox.id || !checkbox.value)
            throw new TypeError('Each checkbox must have an id and value defined')
        }
        return true
      }
    },
    //Optional props
    disabled: Boolean,
    groupId: String,
    groupLabel: String,
    name: String,
    required: Boolean,
    value: Array
  },
  created () {
    //Can't enforce this in the checkboxes validator so this is the earliest we can throw an error
    if (this.checkboxes.length > 1 && (!this.groupId || !this.groupLabel))
      throw new TypeError('Must define group-id and group-label for multiple checkboxes')
  },
  data () {
    return {
      values: this.value || [],
      touched: false
    }
  },
  computed: {
    field () {
      return this.name || this.groupId || this.checkboxes[0].id
    },
    //Wrapper around values so it is propegated through v-model, or 'public' as I've dubbed it
    publicValues: {
      get () {
        return this.values
      },
      set (value) {
        this.touched = true
        this.values = value
        this.$emit('input', this.values)
      }
    },
    validations () {
      let vals = {}
      if (this.required)
        vals.required = true
      return vals
    }
  }
}
</script>

<style scoped>
    
  .baseline {
    align-items: baseline;
    flex: 1;
  }

  .bold {
    font-weight: bold;
  }

  .column {
    display: flex;
    flex-direction: column;
  }

  .flex {
    display:flex;
  }

  .wrap {
    flex-wrap: wrap;
  }

  label {
    padding: 0;
    margin-right: 0.3rem;
  }

  input {
    margin-right: 0.3rem;
  }
</style>
