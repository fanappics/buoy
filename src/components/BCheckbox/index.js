import BCheckbox from './BCheckbox.vue'

/* istanbul ignore next */
BCheckbox.install = function install (Vue) {
  Vue.component(BCheckbox.name, BCheckbox)
}

export default BCheckbox
