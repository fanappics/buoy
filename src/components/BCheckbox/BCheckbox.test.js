import { createLocalVue, shallow } from 'vue-test-utils'
import VeeValidate from 'vee-validate'
import BCheckbox from './BCheckbox'

const localVue = createLocalVue()
localVue.use(VeeValidate)

describe('Single checkbox', () => {
  it('mounted html should match snapshot', () => {
    const wrapper = shallow(BCheckbox, {
      propsData: {
        checkboxes: [{ id: 'checkbox', value: 'Checkbox' }]
      },
      localVue
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('Single disabled checkbox', () => {
  it('mounted html should match snapshot', () => {
    const wrapper = shallow(BCheckbox, {
      propsData: {
        checkboxes: [{ id: 'checkbox', value: 'Checkbox', disabled: true }]
      },
      localVue
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('Single checked checkbox', () => {
  it('mounted html should match snapshot', () => {
    const wrapper = shallow(BCheckbox, {
      propsData: {
        checkboxes: [{ id: 'checkbox', value: 'Checkbox' }],
        value: ['checkbox']
      },
      localVue
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('Multiple checkbox with group label', () => {
  it('mounted html should match snapshot', () => {
    const wrapper = shallow(BCheckbox, {
      propsData: {
        checkboxes: [{ id: 'checkbox-1', value: 'Checkbox', disabled: true }, { id: 'checkbox-2', value: 'Checkbox2' }],
        groupId: 'checkbox-group',
        groupLabel: 'Checkbox Group',
        showGroupLabel: true,
        value: ['checkbox-2']
      },
      localVue
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
