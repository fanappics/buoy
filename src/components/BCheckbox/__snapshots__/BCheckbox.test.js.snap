// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Multiple checkbox with group label mounted html should match snapshot 1`] = `
<div class="b-checkbox b-flex-column">
    <div role="group" aria-labelledby="checkbox-group">
        <label id="group-checkbox-group" class="b-checkbox-group-label">
            Checkbox Group
            <!---->
        </label>
        <div class="b-flex b-flex-wrap">
            <div class="b-flex b-flex-center">
                <input type="checkbox" aria-describedby="" aria-labelledby="group-checkbox-group label-checkbox-1" disabled="disabled" id="checkbox-1" data-vv-as=" " data-vv-name="4" data-vv-scope="4" data-vv-validate-on="input|blur" data-vv-id="4" aria-required="false"
                    aria-invalid="false">
                <label id="label-checkbox-1" for="checkbox-1">
                    Checkbox
                </label>
            </div>
            <div class="b-flex b-flex-center">
                <input type="checkbox" aria-checked="true" aria-describedby="" aria-labelledby="group-checkbox-group label-checkbox-2" id="checkbox-2" data-vv-as=" " data-vv-name="4" data-vv-scope="4" data-vv-validate-on="input|blur" data-vv-id="5" aria-required="false"
                    aria-invalid="false">
                <label id="label-checkbox-2" for="checkbox-2">
                    Checkbox2
                </label>
            </div>
        </div>
    </div>
    <!---->
</div>
`;

exports[`Single checkbox mounted html should match snapshot 1`] = `
<div class="b-checkbox b-flex-column">
    <div class="b-flex b-flex-center">
        <input type="checkbox" aria-describedby="" aria-labelledby="label-checkbox" id="checkbox" data-vv-as=" " data-vv-name="1" data-vv-scope="1" data-vv-validate-on="input|blur" data-vv-id="1" aria-required="false" aria-invalid="false">
        <label id="label-checkbox" for="checkbox">
            Checkbox
            <!---->
        </label>
    </div>
    <!---->
</div>
`;

exports[`Single checked checkbox mounted html should match snapshot 1`] = `
<div class="b-checkbox b-flex-column">
    <div class="b-flex b-flex-center">
        <input type="checkbox" aria-checked="true" aria-describedby="" aria-labelledby="label-checkbox" id="checkbox" data-vv-as=" " data-vv-name="3" data-vv-scope="3" data-vv-validate-on="input|blur" data-vv-id="3" aria-required="false" aria-invalid="false">
        <label id="label-checkbox" for="checkbox">
            Checkbox
            <!---->
        </label>
    </div>
    <!---->
</div>
`;

exports[`Single disabled checkbox mounted html should match snapshot 1`] = `
<div class="b-checkbox b-flex-column">
    <div class="b-flex b-flex-center">
        <input type="checkbox" aria-describedby="" aria-labelledby="label-checkbox" disabled="disabled" id="checkbox" data-vv-as=" " data-vv-name="2" data-vv-scope="2" data-vv-validate-on="input|blur" data-vv-id="2" aria-required="false" aria-invalid="false">
        <label id="label-checkbox" for="checkbox">
            Checkbox
            <!---->
        </label>
    </div>
    <!---->
</div>
`;
